rfd/0150-ssh-connection-resumption.md

```diff
@@ -0,0 +1,129 @@
+
+## Security
+
+The security of the application bytestream tunneled through the resumption protocol is delegated to the protocol used by the application; in this case the application protocol is SSH, which guarantees that all data is encrypted and authenticated, with a secure key exchange. This prevents any attack on the bytestream from becoming an attack on the application data, and the worst that can happen in that regard is that the connection ends up erroring out due to a MAC failure.
+
+Other concerns are strictly about the additional surface provided by the resumption feature itself, however, and will be discussed here.
+
+### Source address information
+
+We propagate the source address of connections entering the Teleport cluster, and the address is stored whenever it's relevant to do so in the audit log. We offer the ability to require that client connections have the same source address as the one used to obtain credentials ([IP pinning](https://goteleport.com/docs/access-controls/guides/ip-pinning/)). As connection resumption intrinsically allows the same connection to roam between different source IP addresses (as the user changes their network configuration, for instance by switching from their wired connection to a mobile hotspot), we shall enforce that the source address of the connection stays the same until the SSH authentication is completed and the user is confirmed to have authenticated with credentials that do not require IP pinning. It's possible that a connection loses its underlying transport before the SSH authentication is complete, and a change in source address at that point will cause the connection to fail; this is acceptable, as the user couldn't have used the connection for anything before authentication anyway.
+
+### Protocol
+
+The protocol in its current form is entirely cleartext; the intended transport for the protocol, however, is secure between the client and the proxy, it's secure between proxies, and it's secure from the proxy to tunneled nodes. An attacker that's able to read traffic between the proxy and direct dial nodes will be able to learn the resumption token, and - with the ability to open connections to the node - the attacker will be able to disrupt the connection. An attacker knowing about a resumption token doesn't have the ability to communicate with the client but, by repeatedly resuming a pre-authentication connection, it's possible for an attacker to take the connection from the intended user and use it as its own. Nothing, however, can be realistically gained from this.
```

Comment:
> repeatedly resuming a pre-authentication connection

One consideration I had is if on reconnect a new resume token should be provided.  Basically the flow would look like this (from the server perspective):
1. exchange offsets
2. generate a new resume token
3. Send missing session data, and as typical with session resumption this data would now be associated to the new resume token (so if another resume was needed before sync, it would be possible as long as the new resume token was received by the client)

The goal would be to allow us to invalidate resume tokens faster.  If a token had been compromised we can now invalidate it after the next resume, and the attacker could not use it to gain continual access to the data stream of the victim.

This seems notably stronger, though arguably whatever mechanism that was used to compromise the resume token initially could likely be leveraged again to get the new tokens.

That said, this strategy is still not as strong as encrypting the resume token.  If the server had a public key from the client we can encrypt the token, enforcing that you must control the private key to be able to resume.

Reply:
What do you suggest? The cost of yet another encrypted tunnel (TLS or similar) in terms of CPU usage might be ok but the latency of establishing (mutual?) TLS adds to what's already a fairly slow and chatty handshake (the SSH one) that we definitely don't want to worsen.